packages/Search/src/details/DTK_DetailsPriorityQueue.hpp

```diff
@@ -26,120 +29,94 @@ template <typename T>
+    KOKKOS_INLINE_FUNCTION Reference top() { return _heap.front(); }
+    KOKKOS_INLINE_FUNCTION ConstReference top() const { return _heap.front(); }
 
-    template <typename... Args>
-    KOKKOS_INLINE_FUNCTION void push( Args &&... args )
+    // Modifiers
+    KOKKOS_INLINE_FUNCTION void push( ValueType const &value )
     {
-        // ensure the heap is not already full
-        assert( _size < _max_size );
-
-        // add the element to the bottom level of the heap
-        IndexType pos = _size;
-        T elem{std::forward<Args>( args )...};
-
-        // perform up-heap operation
-        while ( pos > 0 )
-        {
-            // compare the added element with its parent
-            // if they are in correct order, stop
-            // if not, swap them and continue
-            IndexType const parent = ( pos - 1 ) / 2;
-            if ( !_compare( _heap[parent], elem ) )
-                break;
-            _heap[pos] = _heap[parent];
-            pos = parent;
-        }
-
-        _heap[pos] = elem;
-
-        // update the size of the heap
-        ++_size;
+        _heap.pushBack( value );
+        pushHeap( _heap.data(), _heap.data() + _heap.size(), _compare );
+    }
+    KOKKOS_INLINE_FUNCTION void push( ValueType &&value )
+    {
+        _heap.pushBack( std::move( value ) );
+        pushHeap( _heap.data(), _heap.data() + _heap.size(), _compare );
     }
-
     KOKKOS_INLINE_FUNCTION void pop()
     {
-        // ensure that the heap is not empty
-        assert( _size > 0 );
-
-        // replace the root with the last element on the last level
-        IndexType pos = 0;
-
-        // perform down-heap operation
-        while ( true )
-        {
-            // compare the new root with its children
-            // if they are in the correct order, stop
-            // if not, swap the element with one of its children and continue
-            IndexType const left_child = 2 * pos + 1;
-            IndexType const right_child = 2 * pos + 2;
-            IndexType next_pos = _size - 1;
-            for ( IndexType child : {left_child, right_child} )
-                if ( child < _size - 1 &&
-                     _compare( _heap[next_pos], _heap[child] ) )
-                    next_pos = child;
-            if ( next_pos == _size - 1 )
-                break;
-            _heap[pos] = _heap[next_pos];
-            pos = next_pos;
-        }
-
-        _heap[pos] = _heap[_size - 1];
-
-        // update the size of the heap
-        _size--;
+        popHeap( _heap.data(), _heap.data() + _heap.size(), _compare );
+        _heap.popBack();
     }
-
     // in TreeTraversal::nearestQuery, pop() is often followed by push which is
     // an opportunity for doing a single bubble-down operation instead of paying
     // for both one bubble-down and one bubble-up
     template <typename... Args>
-    KOKKOS_INLINE_FUNCTION void pop_push( Args &&... args )
+    KOKKOS_INLINE_FUNCTION void popPush( Args &&... args )
```

Comment:
Because I changed `push()` to follow stl containers convention.   I have an `emplace()` sitting on my local repo but for some reason I don't know why the radius search got slightly slower when I switched stack insertions to it.  knn timings were unchanged.

Reply:
That said, I chose to forward to `__bubbleDown()` without putting much more thinking into it.  Do you have a better idea?

Reply:
I simply don't know better.